src/shared/kafka-producer-wrapper.ts

```diff
@@ -0,0 +1,90 @@
+import { StatsD } from 'hot-shots'
+import { Producer, ProducerRecord } from 'kafkajs'
+
+import { PluginsServerConfig } from '../types'
+import { timeoutGuard } from './ingestion/utils'
+import { instrumentQuery } from './metrics'
+
+export class KafkaProducerWrapper {
+    /** Kafka producer used for syncing Postgres and ClickHouse person data. */
+    producer: Producer
+    /** StatsD instance used to do instrumentation */
+    statsd: StatsD | undefined
+
+    lastFlushTime: number
+    currentBatch: Array<ProducerRecord>
+    currentBatchSize: number
+
+    flushFrequencyMs: number
+    maxQueueSize: number
+    maxBatchSize: number
+
+    flushInterval: NodeJS.Timeout
+
+    constructor(producer: Producer, statsd: StatsD | undefined, serverConfig: PluginsServerConfig) {
+        this.producer = producer
+        this.statsd = statsd
+
+        this.lastFlushTime = Date.now()
+        this.currentBatch = []
+        this.currentBatchSize = 0
+
+        this.flushFrequencyMs = serverConfig.KAFKA_FLUSH_FREQUENCY_MS
+        this.maxQueueSize = serverConfig.KAFKA_PRODUCER_MAX_QUEUE_SIZE
+        this.maxBatchSize = serverConfig.KAFKA_MAX_MESSAGE_BATCH_SIZE
+
+        this.flushInterval = setInterval(() => this.flush(), this.flushFrequencyMs)
```

Comment:
I guess this interval is good enough, though I think a "fully correct" solution would use a timeout here. It's too easy to get to a situation where the "flush happens 50ms after queueing simply due to the interval boundaries. Though I guess this is a nit and changing it to a different debouncing strategy won't really affect real world performance.

Reply:
The interval was added here (in a previous PR) as an additional protection against data loss. Performance-wise this does not matter and I'd keep our own custom kafka producer impl as lean as possible since we're literally reinventing a wheel in a way I hope to delete some day.